### PR TITLE
Add index_data to data_operations, introduce DataIndex structures

### PR DESCRIPTION
**Pull request by tot0**
Introduce DataIndex configuration and operation.
Preliminary unit and e2e tests added, need work to pass in automation workspaces. 

# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_data_index/data_index.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+
+class CitationRegexSchema(metaclass=PatchedSchemaMeta):
+    match_pattern = fields.Str(
+        required=True,
+        metadata={"description": "Regex to match citation in the citation_url + input file path. e.g. '\\1/\\2'"},
+    )
+    replacement_pattern = fields.Str(
+        required=True,
+        metadata={"description": "Replacement string for citation. e.g. '(.*)/articles/(.*)'"},
```
Comment (gjwoods) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_data_index/data_index.py:33]: Probably want to be consistent with Examples + Component descriptions
'(.*)/articles/(.*)(\\.[^.]+)$'

**Review by gjwoods — APPROVED** [review]
Small suggestions for now or later

File: sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_data_index/data_index.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        metadata={"description": "Input Data to index files from. MLTable type inputs will use `mode: eval_mount`."},
+    )
+    input_glob = fields.Str(
+        required=False,
+        metadata={
+            "description": "Glob pattern to filter files from input_data. If not specified, all files will be indexed."
+        },
+    )
+    chunk_size = fields.Int(
+        load_default=768,
```
[the remark this answers is not in the document]
Comment (tot0) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_data_index/data_index.py:57]: I used same default as in the component, probably better to just have it be fully optional.

File: sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_data_index/data_index.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        metadata={"description": "Maximum number of tokens to put in each chunk."},
+    )
+    chunk_overlap = fields.Int(
```
[the remark this answers is not in the document]
Comment (tot0) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_data_index/data_index.py:60]: chunk_overlap is passed into RCTS config I believe?

**Pull request closed without merging**